ocular/src/keyring.rs

```diff
@@ -0,0 +1,813 @@
+#![warn(unused_qualifications)]
+#![allow(dead_code)]
+// TODO: Remove dead code allowance once these methods are used elsewhere
+
+use bip32::{Mnemonic, PrivateKey};
+use cosmrs::crypto::{secp256k1::SigningKey, PublicKey};
+use rand_core::OsRng;
+use serde::{Deserialize, Serialize};
+use signatory::{
+    pkcs8::der::Document, pkcs8::EncodePrivateKey, pkcs8::LineEnding, FsKeyStore, KeyName,
+};
+use std::collections::HashMap;
+use std::path::Path;
+
+use crate::error::KeyStoreError;
+
+// Constants
+// TODO: Move to independant constants file if reused elsewhere
+const COSMOS_BASE_DERIVATION_PATH: &str = "m/44'/118'/0'/0/0";
+const DEFAULT_FS_KEYSTORE_DIR: &str = "/.ocular/keys";
+
+/// Basic keystore traits that all backends are expected to implement
+pub trait KeyStore {
+    /// Create new key store
+    fn create_key_store(&mut self) -> Result<(), KeyStoreError>;
+
+    /// Check if key store has been initialized.
+    fn key_store_created(&self) -> bool;
+
+    /// Check if key exists under specific name. Will return false if no key is found.
+    fn key_exists(&self, keyname: &KeyName) -> Result<bool, KeyStoreError>;
+
+    /// Add a private key document associated with a key name into the keystore.
+    fn add_key(
+        &self,
+        key_name: &KeyName,
+        encoded_key: pkcs8::PrivateKeyDocument,
+    ) -> Result<(), KeyStoreError>;
+
+    /// Delete key with a given name. If no key exists under name specified an error will be thrown.
+    fn delete_key(&self, key_name: &KeyName) -> Result<(), KeyStoreError>;
+
+    /// Rename key.
+    fn rename_key(&self, current_name: &KeyName, new_name: &KeyName) -> Result<(), KeyStoreError>;
+
+    /// Load PrivateKeyDocumentfrom key store. Will return an error if key DNE under name.
+    fn get_key(&self, key_name: &KeyName) -> Result<pkcs8::PrivateKeyDocument, KeyStoreError>;
+}
+
+/// Mnemonic and private key
+pub struct PrivateKeyOutput {
+    pub mnemonic: Mnemonic,
+    pub private_key: SigningKey,
+}
+
+/// Key name and address in Bech32 (aka segwit) format
+#[derive(Debug)]
+pub struct PublicKeyOutput {
+    /// Name must be unique.
+    pub name: String,
+    pub public_key: PublicKey,
+    pub account: cosmrs::AccountId,
+}
+
+// TODO: As needed, support various record type storage and retrieval, currently it is purely superficial
```

Comment:
Note on this todo; I'm deferring fleshing out various record types to when they're being developed on/with ad hoc in the interest of merging this PR to unblock other issues.